programs/mango-v4/src/state/bank.rs

```diff
@@ -947,12 +951,46 @@ impl Bank {
     ) -> Result<I80F48> {
         require_keys_eq!(self.oracle, *oracle_acc.key());
         let state = oracle::oracle_state_unchecked(oracle_acc, self.mint_decimals)?;
-        state.check_confidence_and_maybe_staleness(
-            &self.oracle,
+        state
+            .check_confidence_and_maybe_staleness(&self.name(), &self.oracle_config, staleness_slot)
+            .with_context(|| oracle_log_context(&state, &self.oracle_config, staleness_slot))?;
+        Ok(state.price)
+    }
+
+    /// Tries to return the primary oracle price, and if there is a confidence or staleness issue returns the fallback oracle price.
+    pub fn oracle_price_with_fallback(
+        &self,
+        oracle_acc: &impl KeyedAccountReader,
+        fallback_oracle_acc_opt: Option<&impl KeyedAccountReader>,
+        staleness_slot: Option<u64>,
+    ) -> Result<I80F48> {
+        require_keys_eq!(self.oracle, *oracle_acc.key());
+        let primary_state = oracle::oracle_state_unchecked(oracle_acc, self.mint_decimals)?;
+        let primary_ok = primary_state.check_confidence_and_maybe_staleness(
+            &self.name(),
             &self.oracle_config,
             staleness_slot,
-        )?;
-        Ok(state.price)
+        );
+        if primary_ok.is_oracle_error() && fallback_oracle_acc_opt.is_some() {
+            let fallback_oracle_acc = fallback_oracle_acc_opt.unwrap();
+            require_keys_eq!(self.fallback_oracle, *fallback_oracle_acc.key());
+            let fallback_state =
+                oracle::oracle_state_unchecked(fallback_oracle_acc, self.mint_decimals)?;
+            let fallback_ok = fallback_state.check_confidence_and_maybe_staleness(
+                &self.name(),
+                &self.oracle_config,
+                staleness_slot,
+            );
+            fallback_ok.with_context(|| {
+                oracle_log_context(&fallback_state, &self.oracle_config, staleness_slot)
```

Comment:
could we also log the primary oracle state here? (just both, we'll know that the first is from the primary)